### PR TITLE
[FLINK-7244] Add parquet table source

### PR DESCRIPTION
**Pull request by HuangZhenQiu**
## What is the purpose of the change

Add projectable and filterable ParquetTableSource based on ParquetInputFormat

## Brief change log

1) Add Parquet table source with expression conversion to Parquet Filters
2) Add unit test and integration test to test the correctness of Fields Projection and Predicate Push

## Verifying this change
  - Added unit test for public functions of ParquetTableSource 
  - Added integration tests for end-to-end test ParquetTableSource in batch mode
 
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetTableSource.java [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+	public static class Builder {
+
+		private String path;
+
+		private MessageType schema;
+
+		private Configuration config;
+
+		private boolean recursive = true;
+
```
Comment (fhueske) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetTableSource.java]: remove empty line

**Comment by HuangZhenQiu** [comment]
@fhueske 
I just resolved your first around of comments. Please add have another look when you have time.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+		assertNotNull(returnType);
+		assertTrue(returnType instanceof RowTypeInfo);
+		RowTypeInfo rowType = (RowTypeInfo) returnType;
+		assertEquals(NESTED_ROW_TYPE, rowType);
+	}
+
+	@Test
+	public void testGetTableSchema() throws Exception {
+		Tuple3<Class<? extends SpecificRecord>, SpecificRecord, Row> nested = TestUtil.getNestedRecordTestData();
+		Path path = TestUtil.createTempParquetFile(tempRoot.getRoot(), TestUtil.NESTED_SCHEMA,
```
Comment (fhueske) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java]: No need to create a temp file

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+		ParquetTableSource parquetTableSource = createNestedTestParquetTableSource();
+
+		ParquetTableSource projected = (ParquetTableSource) parquetTableSource.projectFields(new int[] {2, 4, 6});
+
+		// ensure copy is returned
+		assertTrue(projected != parquetTableSource);
+
+		// ensure table schema is the same
+		assertEquals(parquetTableSource.getTableSchema(), projected.getTableSchema());
+
```
Comment (fhueske) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java]: Check that the source description is different (`!parquetTableSource.explainSource.equals(projected.explainSource())`)

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetTableSource.java [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+	}
+
+	@Override
+	public TableSchema getTableSchema() {
+		return tableSchema;
+	}
+
+	@Override
+	public String explainSource() {
+		return "ParquetFile[path=" + path + ", schema=" + parquetSchema + ", filter=" + predicateString() + "]";
```
Comment (fhueske) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetTableSource.java]: Include the projected fields in the source description. Flink checks whether the project pushdown worked by comparing the descriptions. If both are the same, the rule is not applied.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+		assertEquals(expected, arguments.getValue());
+
+		assertTrue(spyTableSouce.isFilterPushedDown());
+		assertFalse(parquetTableSource.isFilterPushedDown());
+	}
+
+	private ParquetTableSource createNestedTestParquetTableSource() throws Exception {
+		Tuple3<Class<? extends SpecificRecord>, SpecificRecord, Row> nested = TestUtil.getNestedRecordTestData();
+		Path path = TestUtil.createTempParquetFile(tempRoot.getRoot(), TestUtil.NESTED_SCHEMA,
```
Comment (fhueske) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java]: No need to create temp files here.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+		spyPTS.getDataSet(env);
+		verify(mockIF).selectFields(eq(new String[] {fieldNames[2], fieldNames[4], fieldNames[6]}));
+	}
+
+	@Test
+	public void testFieldsFilter() throws Exception {
+		ParquetTableSource parquetTableSource = createNestedTestParquetTableSource();
+
+		// expressions for supported predicates
+		Expression exp1 = new GreaterThan(
```
Comment (fhueske) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java]: please check the inverted case (literal op column) as well

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+		// apply predict on TableSource
+		ParquetTableSource projected = (ParquetTableSource) parquetTableSource.applyPredicate(exps);
+
+		// ensure copy is retured
+		assertTrue(parquetTableSource != projected);
+
+		// ensure table schema is identical
+		assertEquals(parquetTableSource.getTableSchema(), projected.getTableSchema());
+
```
Comment (fhueske) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java]: Check that the pushed down predicates were removed from `exps`.

**Comment by HuangZhenQiu** [comment]
@fhueske I resolve all of the comments, except the last one. Not sure why "the pushed down predicates should be removed from exps.

**Comment by fhueske** [comment]
Hi @HuangZhenQiu, the predicates should be removed to let the optimizer know that these are handled by the table source. If they are not removed, the optimizer will generate code to evaluate the predicates which was already done by the table source.

I'll check the PR now.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetTableSource.java [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+ *   .schema(messageType)
+ *   .build();
+ *
+ * tEnv.registerTableSource("parquetTable", orcSrc);
+ * Table res = tableEnv.sqlQuery("SELECT * FROM parquetTable");
+ * }
+ * </pre>
+ */
+public class ParquetTableSource
+	implements BatchTableSource<Row>, FilterableTableSource<Row>, ProjectableTableSource<Row> {
```
[the remark this answers is not in the document]
Comment (fhueske) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetTableSource.java]: OK, seems that we would need to add a dependency on `flink-streaming` for this. I'll keep it as it is for now.

**Pull request closed without merging**